### ISSUE TITLE
Multiple Requests Causes User Logout

### ISSUE THREAD

**Issue by camconn** [niemal/uman#1]
Whenever a user makes multiple requests (for example, while fetching the image or CSS assets for a page), uman logs the user out and forces them to log back in.

**Comment by camconn** [comment]
To Elaborate:

In [this branch](https://github.com/camconn/shim/tree/thumb) of Shim, there are multiple thumbnails on a single page. Fetching the page a few times causes for uman to indicate that the user isn't logged in. This also isn't specific to thumbnails, because you can replicate this bug by reloading any page in shim enough times.

I suspect that this is caused by a race condition related to maps in Go. I'm testing this code now.

This isn't a critical bug, but it _is_ annoying for end-users and developers.

**Comment by camconn** [comment]
It is related to data races. Seems to be caused whenever two separate goroutines refer to the same session.

```
==================
WARNING: DATA RACE
Write by goroutine 8:
  runtime.mapassign1()
      /home/cam/go/go1.6/src/runtime/hashmap.go:429 +0x0
  github.com/niemal/uman.(*UserManager).GetHTTPSession()
      /home/cam/go/local/src/github.com/niemal/uman/uman.go:324 +0x327
  main.(*loginHandler).authHandler.func1()
      /home/cam/Dev/shim/handlers.go:61 +0x94
  net/http.HandlerFunc.ServeHTTP()
      /home/cam/go/go1.6/src/net/http/server.go:1618 +0x47
  main.loggingHandler.func1()
      /home/cam/Dev/shim/handlers.go:32 +0x21e
  net/http.HandlerFunc.ServeHTTP()
      /home/cam/go/go1.6/src/net/http/server.go:1618 +0x47
  net/http.(*ServeMux).ServeHTTP()
      /home/cam/go/go1.6/src/net/http/server.go:1910 +0x212
  net/http.serverHandler.ServeHTTP()
      /home/cam/go/go1.6/src/net/http/server.go:2081 +0x206
  net/http.(*conn).serve()
      /home/cam/go/go1.6/src/net/http/server.go:1472 +0x1565

Previous read by goroutine 6:
  runtime.mapiterinit()
      /home/cam/go/go1.6/src/runtime/hashmap.go:607 +0x0
  github.com/niemal/uman.(*UserManager).CheckSessions()
      /home/cam/go/local/src/github.com/niemal/uman/uman.go:257 +0x8a

Goroutine 8 (running) created at:
  net/http.(*Server).Serve()
      /home/cam/go/go1.6/src/net/http/server.go:2137 +0x4d1
  net/http.(*Server).ListenAndServe()
      /home/cam/go/go1.6/src/net/http/server.go:2098 +0x174
  main.main()
      /home/cam/Dev/shim/main.go:123 +0x1c07

Goroutine 6 (running) created at:
  github.com/niemal/uman.New()
      /home/cam/go/local/src/github.com/niemal/uman/uman.go:104 +0x17b
  main.main()
      /home/cam/Dev/shim/main.go:67 +0x212c
==================
2016/03/16 11:58:12 Not logged in! Redirecting to /login/?redirect=%2Fadmin%2F&warn=yes
2016/03/16 11:58:12 [GET] "/login/?redirect=%2Fadmin%2F&warn=yes"
2016/03/16 11:58:12 [GET] "css/bulma-0.0.15.min.css"
2016/03/16 11:58:12 [GET] "css/layout.css"
2016/03/16 11:58:12 [GET] "css/global.css"
2016/03/16 11:58:13 [GET] "fontello/css/elusive_shim.css"
2016/03/16 11:58:13 [GET] "fontello/font/elusive_shim.woff2?38573169"
2016/03/16 11:58:15 [POST] "/login/"
[UserManager]: User[root] has logged in.
2016/03/16 11:58:16 Redirecting to /admin/
2016/03/16 11:58:16 [GET] "/admin/"
2016/03/16 11:58:19 [POST] "/admin/"
Opening config at /home/cam/Dev/shim/sites/mysite/config.toml
2016/03/16 11:58:19 Temporarily http://127.0.0.1:8080/preview/
2016/03/16 11:58:20 Now https://www.camconn.cc/
==================
WARNING: DATA RACE
Read by goroutine 6:
  github.com/niemal/uman.(*UserManager).CheckSessions()
      /home/cam/go/local/src/github.com/niemal/uman/uman.go:258 +0x150

Previous write by goroutine 8:
  [failed to restore the stack]

Goroutine 6 (running) created at:
  github.com/niemal/uman.New()
      /home/cam/go/local/src/github.com/niemal/uman/uman.go:104 +0x17b
  main.main()
      /home/cam/Dev/shim/main.go:67 +0x212c

Goroutine 8 (running) created at:
  net/http.(*Server).Serve()
      /home/cam/go/go1.6/src/net/http/server.go:2137 +0x4d1
  net/http.(*Server).ListenAndServe()
      /home/cam/go/go1.6/src/net/http/server.go:2098 +0x174
  main.main()
      /home/cam/Dev/shim/main.go:123 +0x1c07
==================
==================
WARNING: DATA RACE
Read by goroutine 6:
  github.com/niemal/uman.(*UserManager).CheckSessions()
      /home/cam/go/local/src/github.com/niemal/uman/uman.go:258 +0x163

Previous write by goroutine 8:
  [failed to restore the stack]

Goroutine 6 (running) created at:
  github.com/niemal/uman.New()
      /home/cam/go/local/src/github.com/niemal/uman/uman.go:104 +0x17b
  main.main()
      /home/cam/Dev/shim/main.go:67 +0x212c

Goroutine 8 (running) created at:
  net/http.(*Server).Serve()
      /home/cam/go/go1.6/src/net/http/server.go:2137 +0x4d1
  net/http.(*Server).ListenAndServe()
      /home/cam/go/go1.6/src/net/http/server.go:2098 +0x174
  main.main()
      /home/cam/Dev/shim/main.go:123 +0x1c07
==================
```

**Comment by camconn** [comment]
Unfortunately, this bug is inherent to uman's architecture because uman uses a central session store. Perhaps this could be mitigated by storing _all_ session information with the user?

Doing something similar to what [securecookie](https://github.com/gorilla/securecookie) does is worth exploring. Note that while this _does_ open up the cookie to user tampering, it is nearly the same risk as keeping a cookie associated with a random session id.

**Comment by niemal** [comment]
It appears that this is a thread concurrency issue and an attempt to fix this issue is a25fde4. Let me know if that works.

**Comment by camconn** [comment]
Issue is resolved with recent commits. Thanks
